### ISSUE TITLE
Risky octal shouldn't recommend `0` prefixing the permissions.

### ISSUE THREAD

**Issue by jcgruenhage** [ansible/ansible-lint#2965]
##### Summary

`yaml` is a notoriously complicated format, with different pitfalls depending on which version you're looking at. What Ansible Lint recommends for permissions only works because the underlying `yaml` parser does not implement the latest version (which has been out for over a decade). Octal numbers in yaml specification >= 1.2.2 work by prefixing `0o` and not `0`.

My suggestion here would be to mandate strings and completely disallow int literals. This avoids the ambiguity between encodings of different yaml versions. Mandating `0o` can't work, because that's not supported by ansible.

##### Issue Type

- Bug Report

##### OS / ENVIRONMENT

not applicable

##### STEPS TO REPRODUCE

not applicable

##### Desired Behavior

Don't recommend setting values that will break when the yaml parser is eventually upgraded to the new version.

##### Actual Behavior

It recommends prefixing the int with an `0`, which will break eventually.

**Comment by ssbarnea** [comment]
I checked and while Ansible supports the new `0o` prefixes documented by [YAML 1.2.2 specification](https://yaml.org/spec/1.2.2/ext/changes/), that is apparently not documented anywhere.


- [Ansible YAML Syntax page](https://docs.ansible.com/ansible/latest/reference_appendices/YAMLSyntax.html) does not contain any documentation about octals and it should be updated to include that ASAP
- Core modules accepting octals (usually `mode` argument), still advertise the old octals and we need to fix them.
- We will update the linter to recommend the modern octals, so people will have plenty of time to migrate to more future proof values.

**Comment by ssbarnea** [comment]
@jcgruenhage I am quite curious about your statement related to "0o can't work, because that's not supported by ansible." -- which based on my own tests prove to be false. Yes, it is not documented, but when trying to use `0o` prefix with file module, I got perfect results. Maybe some 3rd party modules might fail to use the correct ansible methods for processing the octals, but Ansible itself seems to work fine with new ones.

**Comment by jcgruenhage** [comment]
https://github.com/ansible/ansible/blob/fa382670dc35d89652c03c3d7f7911533a36b29d/lib/ansible/module_utils/basic.py#L897 is the line that parses the strings, for which apparently both `0o755` and `0755` turn into `493` in decimal, so that solves that bit of mystery.

**Comment by sivel** [comment]
I should also note that jinja2 native resolves the issues as well, and we hope to switch to that being the default soon.

**Comment by jcgruenhage** [comment]
@sivel can you elaborate on that last bit? How is Jinja related to this quirk in yaml parsing?

**Comment by nitzmahone** [comment]
Yeah, Ansible doesn't "really" support the `0o` prefix. Someone long ago added some module code in a few places to catch obvious (and specific) cases of octal file-mode value munging. However, it's a hack that's only manually applied down at the module level by certain modules, so anything else will still see the raw value as things stand today. eg:

```yaml
- hosts: localhost
  gather_facts: no
  vars:
    old_octal: 0123
    new_octal: 0o123
    intvalue: 1 
  tasks:
  - debug:
      msg: jinja says {{ new_octal + intvalue }}
  - debug:
      msg: jinja says {{ old_octal + intvalue }}

```

The `new_octal` case blows up on current Ansible because PyYAML just gives us back a string.

We could adjust core's usage of PyYAML to eliminate the most egregious YAML 1.1 typing "surprises" (even without true YAML 1.2 support in PyYAML yet) with some combination of the following:

1) Add support for the 1.2 octal int format (`0o` -prefixed) to the implicit `int` resolver and scalar constructor, so that format is fully-supported.
2) Remove support for the 1.1 octal int format (`0` -prefixed) from only the scalar constructor (so `0`-prefixed ints would resolve as they do in 1.2, ie, just a 0-padded base10 `int` - without also *removing* any octal-munging code from core (and how many 3rd party modules have emulated this?), this would probably cause problems.
3) Remove support for the 1.1 octal int format from *only* the implicit `int` resolver- this causes all `0` -prefixed ints to resolve to strings (best for octal modes, sucks if you actually want octal numeric support for any other reason).
4) Remove support for 1.1 `bool` values (ie, anything but `true`|`false`)
5) (most drastic) Remove support for *all* YAML scalar types (eg, every scalar is a string- rely only on Jinja and Ansible's own type coercion).

All of these are surprisingly easy to implement in core itself by just setting up the PyYAML loader differently- I've gotten them all working. #&#x2060;5 would probably cause more trouble with Jinja expressions referencing numeric scalars than it's worth, but it would also neatly solve the octal mode munging problem (since just eliminating `0` prefixed octal still doesn't do that, it just makes it more predictable).

**Comment by sivel** [comment]
> @sivel can you elaborate on that last bit? How is Jinja related to this quirk in yaml parsing?

This problem is somewhat complex, but the "danger" in using octals, largely comes with using a loop, or references to some other variable.

Here is an example:

```yaml
- file:
    path: '{{ item.path }}'
    mode: '{{ item.mode }}'
  loop:
    - path: /foo
      mode: 0755
```

What happens, is that YAML parses `0755` into an int of `493` at parsing time.  Now, due to the historical implementation of jinja2, a single templating action such as `{{ item.mode }}` can only produce a python bool, dict, list, set, or a str.  It cannot produce an `int`.  As such, `mode: '{{ item.mode }}'` really is equivalent to `mode: !!str 493` (I used `!!str` here to just to not confuse with quotes).

As such, when we send the str `'493'` through to `os.chmod`, it's seen as a str, and then the python internals would assume the `493` is actually `0493` and not `0755`.

jinja2 native, would allow the integer of `493` to remain an integer, and `os.chmod` wouldn't need to do any conversions of the data, and thus you would end up with a mode of `0755` being applied in the end.

---
However, if we step back to just the original concern here:

> Don't recommend setting values that will break when the yaml parser is eventually upgraded to the new version

See the other response here https://github.com/ansible/ansible-lint/issues/2965#issuecomment-1414400161

Ultimately, recommending we use `0o755` format isn't really the right thing to just do, because YAML 1.1 and 1.2 treat them completely different.

Ideally as I mention, the most flexible solution is for users to make sure their octals are strings, by use of quotes or the `!!str` tag, to bypass any YAML parsing behaviors of octals.

---

fwiw, I do know of several oddballs that are specifically providing `int` values in `mode` values, and not octals at all.

**Comment by ssbarnea** [comment]
I am wondering if ruamel.yaml is able to be tuned to report us if a string has explicit quotes or not in YAML. I looked and by default it does not provide this information.

If we want to enforce use of strings, we need a way to identify explicit strings in a way that works well for all versions of the specification.

**Comment by nitzmahone** [comment]
Ruamel's certainly more likely to be able to give you that than PyYAML probably ever will, since it already supports sidecar smuggling of input formatting for round-trips. The parser does store the quote style on the `style` attribute of the node (at least PyYAML's does, and Ruamel's was built from that), so it should be available if there's a place to stash it in Ruamel's formatting cache post-construction.

**Comment by samccann** [comment]
FWIW - the yaml syntax docs page is one of our top pages so if y'all think it needs updates, please do open a PR.
